### PR TITLE
fix depends for array builder

### PR DESCRIPTION
**Pull request by rhasselle-oddball**
## Summary

- Fix `depends` for array builder and add `index` property

```
depends: (formData, index) => ... 
```

- Fix forward navigation not respecting newly updated values for all arrays

## Related issue(s)

- _Link to ticket created in va.gov-team repo_
department-of-veterans-affairs/va.gov-team-forms#1800
department-of-veterans-affairs/va.gov-team-forms#1888

## Testing done

- Unit, cypress, regression, browser

## Screenshots

n/a

## What areas of the site does it impact?

array builder `depends`.

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Review by mkerns1 — DISMISSED** [review]
This looks good to me! Thanks so much for adding this in!

**Review by Midge-dev — DISMISSED** [review]

File: src/platform/forms-system/src/js/patterns/array-builder/README.md [hunk 1]
```diff
@@ -444,7 +445,27 @@ const options = {
...
+    path: 'array-multiple-page-builder-summary',
+    uiSchema: employersSummaryPage.uiSchema,
+    schema: employersSummaryPage.schema,
+  }),
+  multiPageBuilderStepOne: pageBuilder.itemPage({
+    title: 'Employer name and address',
+    path: 'array-multiple-page-builder/:index/name-and-address',
+    uiSchema: employersPageNameAndAddressPage.uiSchema,
+    schema: employersPageNameAndAddressPage.schema,
+    depends: (formData, index) => formData.employers?.[index]?.type === 'Military',
```
Comment (ToddWebDev) [inline src/platform/forms-system/src/js/patterns/array-builder/README.md:463]: If `formData.employers` is an array in this example, then shouldn't the syntax be this to access the type property:
```suggestion
    depends: (formData, index) => formData.employers[index]?.type === 'Military',
```

**Review by powellkerry — DISMISSED** [review]

File: src/platform/forms-system/src/js/patterns/array-builder/arrayBuilder.jsx [hunk 2]
```diff
@@ -336,30 +336,50 @@ export function arrayBuilderPages(options, pageBuilderCallback) {
...
+  const navForwardFinishedItem = ({ goPath, urlParams, pathname, index }) => {
     let path = summaryPath;
     if (urlParams?.edit || (urlParams?.add && urlParams?.review)) {
-      const index = getArrayIndexFromPathName(pathname);
+      const foundIndex = getArrayIndexFromPathName(pathname);
       const basePath = urlParams?.review ? reviewPath : summaryPath;
       path = createArrayBuilderUpdatedPath({
         basePath,
-        index,
+        index: foundIndex == null ? index : foundIndex,
```
Comment (Midge-dev) [inline src/platform/forms-system/src/js/patterns/array-builder/arrayBuilder.jsx:382]: Just curious if there is a reason for `==` vs `===`?

Comment (rhasselle-oddball) [reply src/platform/forms-system/src/js/patterns/array-builder/arrayBuilder.jsx:382]: Good catch, yes this was actually intentional. One of the rare cases I use ==, to handle both null and undefined.

**Review by ToddWebDev — DISMISSED** [review]
Works well for me!

**Review by williamphelps13 — COMMENTED** [review]
Just found a potential bug. I had the page that was shown conditionally as a middle page of the loop (that is what the above video demos). With a recent request the conditional page became the last page of the loop. On the summary page I am now not able to click the "no" option and and continue on. The following video shows the conditional page as a middle page first then as the last page:

https://github.com/user-attachments/assets/929adfb2-b257-42ec-b0d2-c14b4e2a366f

Here is my branch with your changes merged in and the conditional page as the last in the loop (select the "Cherry" demo when in localhost): conditions-639-refine-prototypes

**Comment by rhasselle-oddball** [comment]
Thanks @Midge-dev for helping with the tests on 686

Status update:   Working on bug that @williamphelps13 found.  Need to fix that before merging.

File: src/platform/forms-system/src/js/patterns/array-builder/arrayBuilder.jsx [hunk 3]
```diff
@@ -336,63 +336,96 @@ export function arrayBuilderPages(options, pageBuilderCallback) {
...
+    const nextIndex = formData[arrayPath]?.length || 0;
+
+    const path = createArrayBuilderItemAddPath({
+      path: getFirstItemPagePath(formData, nextIndex),
+      index: nextIndex,
+    });
+    goPath(path);
+  };
+
+  const navForwardSummaryGoNextChapter = ({ formData, goPath, pageList }) => {
```
Comment (rhasselle-oddball) [inline src/platform/forms-system/src/js/patterns/array-builder/arrayBuilder.jsx:400]: This part is new in the latest changes, which fixes the bugs that @williamphelps13 had discovered. (The index was sometimes wrong in a previous commit for this PR)

File: src/applications/disability-benefits/686c-674-v2/config/form.js [hunk 4]
```diff
@@ -593,26 +593,26 @@ export const formConfig = {
...
-            onNavForward: ({
-              formData,
-              pathname,
-              urlParams,
-              goPath,
-              goNextPath,
-            }) => {
-              const index = getArrayIndexFromPathName(pathname);
-              const urlParamsString = stringifyUrlParams(urlParams) || '';
+            //   onNavForward: ({
```
Comment (rhasselle-oddball) [inline src/applications/disability-benefits/686c-674-v2/config/form.js:596]: I commented out some additional cases in the latest commit. Is this OK @Midge-dev ?

File: src/applications/simple-forms/mock-simple-forms-patterns/tests/e2e/mock-array-builder-depends.cypress.spec.js [hunk 5]
```diff
@@ -0,0 +1,289 @@
+import path from 'path';
```
Comment (rhasselle-oddball) [inline src/applications/simple-forms/mock-simple-forms-patterns/tests/e2e/mock-array-builder-depends.cypress.spec.js:1]: Added new cypress test in latest commit.

File: src/applications/disability-benefits/686c-674-v2/config/form.js (hunk 4 above)
Comment (Midge-dev) [reply src/applications/disability-benefits/686c-674-v2/config/form.js:596]: Yes that's fine - we plan to rework all of this during clean up. Tyvm!

File: src/applications/simple-forms/21-0845/tests/e2e/0845-auth-disclose.cypress.spec.js [hunk 6]
```diff
@@ -42,6 +42,14 @@ const testConfig = createTestConfig(
           });
         });
       },
+      'veteran-identification-information': ({ afterHook }) => {
```
Comment (rhasselle-oddball) [inline src/applications/simple-forms/21-0845/tests/e2e/0845-auth-disclose.cypress.spec.js:45]: included this additional unrelated change - because it was blocking PR

**Review by rhasselle-oddball — COMMENTED** [review]
PR ready for review

**Review by Midge-dev — DISMISSED** [review]

**Review by williamphelps13 — COMMENTED** [review]
Previous bug: when dependent page is the last page and its the last item added the user could not continue on from the list loop
This is now resolved.

Potential new issue: the recent changes appear to have changed formData to contain all formData for onNavForward when it previously only contained the itemData.

This may be an enhancement *(in the long term) because it seems desirable that formData always contain all formData and never just contain itemData. *Though in this PR it will cause regressions in some forms.

**Comment by rhasselle-oddball** [comment]
> Previous bug: when dependent page is the last page and its the last item added the user could not continue on from the list loop This is now resolved.
> 
> Potential new issue: the recent changes appear to have changed formData to contain all formData for onNavForward when it previously only contained the itemData.
> 
> This may be an enhancement actually because it seems desirable that formData always contain all formData and never just contain itemData.

Appreciate the thorough testing @williamphelps13 - I'll spend some more time on this, to make sure the `formData` discrepancies don't cause regressions in existing forms.

**Review by williamphelps13 — APPROVED** [review]
Changes are working well in the 2 `new-conditions` multi-page list and loops that use onNavForward. If I convert them to depends they are also working well.

**Review by ToddWebDev — APPROVED** [review]

**Review by ndsprinkle — APPROVED** [review]

File: src/applications/disability-benefits/686c-674-v2/config/form.js [hunk 7]
```diff
@@ -593,6 +593,7 @@ export const formConfig = {
             schema: studentEducationBenefitsPage.schema,
             depends: formData =>
               isChapterFieldRequired(formData, TASK_KEYS.report674),
+            // TODO: use depends: (formData, index) instead on the dynamic page.
```
Comment (Midge-dev) [inline src/applications/disability-benefits/686c-674-v2/config/form.js:596]: Solid. Thank you!

**Review by Midge-dev — APPROVED** [review]

**Review by rsmithadhoc — APPROVED** [review]
